### PR TITLE
FileLoader now accepts Promise instead of a File instance

### LINKED ISSUE #87: It should be possible to use file loader with asynchronous file fetching
Extracted from https://github.com/ckeditor/ckeditor5-paste-from-office/issues/44#issuecomment-443200544:

> We realised that this code:
> 
> https://github.com/ckeditor/ckeditor5-image/blob/14dca372ab3e57e8aa4a8eead5d55fffe3143d5c/src/imageupload/imageuploadediting.js#L101-L137
> 
> does not have to stop the event. It can create loader instances synchronously and set the `uploadId` attribute synchronously. However, it does not have `file` instances synchronously – that must remain asynchronous. Which means that we need to be able to create loader instances like this:
> 
> ```js
> const loader = fileRepository.createLoader( new Promise( resolve => {
>     resolve( file );
> } ) );
> 
> loader.file; // -> null
> loader.status; // -> 'uninitialized'
> loader.read(); // errror!
> loader.id; // -> 'idoftheloader'
> 
> // promise gets resolved by the `ImageUploadEngine` once `file` instance is returned by the native > File API or whatever else...
> 
> loader.file // -> file instance
> loader.status; // -> 'idle'
> loader.id; // -> 'idoftheloader'
> ```
> 
> It's important that `loader.id` is accessible from the first moment (synchronously) so the main logic of that listener can stay synchronous.

### PR DESCRIPTION
**Pull request by f1ames**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: `FileLoader` now accepts `Promise` instead of a `File` instance. Closes #87.

BREAKING CHANGE: The `FileLoader.file` property was changed to a getter which returns a native `Promise` instance instead of a `File` instance. The returned promise resolves to a `File` instance.

---

### Additional information

See #87 for technical explanations what was changed.

This change also required some updates in `ckeditor5-image` and `ckeditor5-easy-image` plugins (because of a change in how `loader.file` works and it's async nature):

* Related `ckeditor5-image` PR: https://github.com/ckeditor/ckeditor5-image/pull/258.
* Related `ckeditor5-easy-image` PR: https://github.com/ckeditor/ckeditor5-easy-image/pull/22.
* Related `ckeditor5-adapter-ckfinder` PR: https://github.com/ckeditor/ckeditor5-adapter-ckfinder/pull/13.

### CODE REVIEW

**Comment by ma2ciek** [comment]
All new promises in API docs can have the `Promise.<File>` type.

**Comment by jodator** [comment]
> All new promises in API docs can have the `Promise.<File>` type.

thanks for tip @ma2ciek

File: src/filerepository.js [hunk 1]
```diff
@@ -354,6 +373,25 @@ class FileLoader {
...
+	 * Returns a `Promise` which resolves to a `File` instance associated with this file loader.
+	 *
+	 * @type {Promise}
+	 */
+	get file() {
+		if ( !this._filePromiseWrapper ) {
+			// Loader was destroyed, return promise which resolves to null.
+			return new Promise( resolve => resolve( null ) );
+		} else {
+			// The `loader.file.then( ... )` can be called, then `loader._destroy()` (which is sync) and then
```
Comment (jodator) [inline src/filerepository.js]: I got lost :) Could you rephrase it a bit?

File: src/filerepository.js [hunk 2]
```diff
@@ -354,6 +373,25 @@ class FileLoader {
...
+			// `loader.file.then( ... )` will be resolved. In such case, even after `destroy` call the promise will
+			// return `File` instance so we use nested promises which handles this case.
+			return this._filePromiseWrapper.promise.then( file => {
+				return new Promise( resolve => resolve( this._filePromiseWrapper ? file : null ) );
```
Comment (jodator) [inline src/filerepository.js]: Similarly to `if()` this can be shortened:
```js
return this._filePromiseWrapper.promise
    .then( file => this._filePromiseWrapper ? file : null );
```
as the promise from `this._filePromiseWrapper` will be used to create a promise chain and the returned value will passed to the `.then()` call on `loader.file.then()`.

File: src/filerepository.js [hunk 3]
```diff
@@ -354,6 +373,25 @@ class FileLoader {
...
 
+	/**
+	 * Returns a `Promise` which resolves to a `File` instance associated with this file loader.
+	 *
+	 * @type {Promise}
+	 */
+	get file() {
+		if ( !this._filePromiseWrapper ) {
+			// Loader was destroyed, return promise which resolves to null.
+			return new Promise( resolve => resolve( null ) );
```
Comment (jodator) [inline src/filerepository.js]: Could be:
```js
return Promise.resolve( null )
```

File: src/filerepository.js [hunk 4]
```diff
@@ -265,12 +283,13 @@ class FileLoader {
 		this.id = uid();
 
 		/**
-		 * A `File` instance associated with this file loader.
+		 * Additional wrapper over a file promise passed to this loader.
```
Comment (jodator) [inline src/filerepository.js]: The constructor of `FileLoader` has outdated docs for `@param {File}` which should be `@param {Promise.<File>}`.

Comment (jodator) [reply src/filerepository.js]: I cannot comment the code that wasn't changed hence the location of this comment.

File: src/filerepository.js [hunk 5]
```diff
@@ -265,12 +283,13 @@ class FileLoader {
...
 		 *
-		 * @readonly
-		 * @member {File}
+		 * @see #_createFilePromiseWrapper
+		 * @private
+		 * @member {Object} wrapper
```
Comment (jodator) [inline src/filerepository.js]: Might be nice to create `@typedef` for the wrapper to document it instead of just `Object`.

File: src/filerepository.js (hunk 4 above)
Comment (jodator) [reply src/filerepository.js]: Also the param name to be updated also

File: src/filerepository.js [hunk 6]
```diff
@@ -174,10 +176,20 @@ export default class FileRepository extends Plugin {
 			return null;
 		}
 
-		const loader = new FileLoader( file );
+		const filePromise = fileOrPromise instanceof Promise ? fileOrPromise : new Promise( resolve => resolve( fileOrPromise ) );
```
Comment (jodator) [inline src/filerepository.js]: I'd shorten that using chainability of promises
```js
const loader = new FileLoader( Promise.resolve( fileOrPromise ) );
```

If value is returned it will be wrapped in promise. If you return a promise then it will be simply returned and work as intended.

File: src/filerepository.js [hunk 7]
```diff
@@ -174,10 +176,20 @@ export default class FileRepository extends Plugin {
...
+		const loader = new FileLoader( filePromise );
+
 		loader._adapter = this.createUploadAdapter( loader );
```
Comment (jodator) [inline src/filerepository.js]: I've just stumbled upon this and didn't realized it earlier. The docs for `FileLoader`'s constructor are rubbish then.

It's a chicken and egg problem as Loader needs adapter and creating adapter requires loader :D

Please remove it from constructor's docs and maybe create some privete method to hide this obscurity :)

I'd leave here:
```js
const loader = this._initLoader( fileOrPromise );
```
so all the above code would go the the `initLoader` method. Also please add some coment above `loader._adaper = ` part there as it should be explained why this trick ( AFAICS the `this.createUploadAdapter()` expects the loader instance.

File: src/filerepository.js [hunk 8]
```diff
@@ -120,32 +128,26 @@ export default class FileRepository extends Plugin {
 	}
 
 	/**
-	 * Returns the loader associated with specified file.
+	 * Returns the loader associated with specified file or promise.
 	 *
 	 * To get loader by id use `fileRepository.loaders.get( id )`.
 	 *
-	 * @param {File} file Native file handle.
+	 * @param {File|Promise} fileOrPromise Native file or promise handle.
```
Comment (jodator) [inline src/filerepository.js]: As Maciek wrote - you can use `Promise.<File>` to represent this. Also please check other occurrences of `Promise` in the docs to change them to `Promise.<File>`.

File: src/filerepository.js (hunk 4 above)
Comment (f1ames) [reply src/filerepository.js]: 👍

File: src/filerepository.js (hunk 7 above)
Comment (f1ames) [reply src/filerepository.js]: Didn't notice it too. 😂 🐔 Maybe instead of doing `loader._adapter` which looks bad, we could pass `createUploadAdapter` function as a second param of a `FileLoader` constructor:

```js
class FileLoader {
	constructor( filePromise, createAdapterFn ) {
        this._adapter = createAdapterFn( this );
    }
}
```

it makes it a little more complicated and coupled, but as `FileRepository` creates adapter only to "pass" it to the loader, maybe loader should create adapter itself? I'm not sure though, WDYT?

File: src/filerepository.js [hunk 9]
```diff
@@ -376,19 +372,22 @@ class FileLoader {
 	/**
 	 * Returns a `Promise` which resolves to a `File` instance associated with this file loader.
```
[the remark this answers is not in the document]
Comment (ma2ciek) [reply src/filerepository.js]: Oh, it's not your code. Sorry for that 😄

[the remark this answers is not in the document]
Comment (f1ames) [reply src/filerepository.js]: I think it was initially created by me, not sure 🤔 😄 Anyway, I corrected it 👍

**Comment by f1ames** [comment]
> ps.: Shouldn't we also update CKFinder upload adapter?

Looking at this line:

```js
data.append( 'upload', this.loader.file );
```

I think we should 😄  Strangely, all unit tests are passing 🤔

**Comment by f1ames** [comment]
I have rechecked https://github.com/ckeditor/ckeditor5-easy-image/pull/22 and https://github.com/ckeditor/ckeditor5-image/pull/258 branches after refactor and updated `adapter-ckfinder` package too - https://github.com/ckeditor/ckeditor5-adapter-ckfinder/pull/13.
